### ISSUE TITLE
What should be the `codecs` parameter of the single track approach?

### ISSUE THREAD

**Issue by cconcolato** [AOMediaCodec/iac#22]
Current audio codecs have a MIME `codecs` parameter of the following forms:
- `Opus`
- `mp4a.40.2`

We should define what the MIME `codecs` parameter should be for the single track case. The purpose of the `codecs` parameter is for a receiver to determine if it can process the track without having to download anything (no header, no initialization segment, ...).

Do we expect that all IAC decoders will be able to process every file or do we expect that some files won't be processable? In the latter case, the `codecs` should contain something. Also, we should recursively convey the underlying `codecs` parameter. For example something like: `aiac.<IAC-specific-needs>.mp4a.40.2` or `aiac.<IAC-specific-needs>.Opus`.

**Comment by sunghee-hwang** [comment]
Thank you for the information. Based on your comments, we may need to categorize it further as follows:
1. IAC decoders to support Ambisonics: In this case, they may or may not be able to process a file dependent on the order of Ambisonics. So we may need something like "ambix#" or "a#" for <IAC-Specific-needs>. a0: no Ambisonics, a1: FOA, a2: SOA etc..
2. IAC decoders to support Channel Audio: In this case, they may or may not be able to process every layouts of the file: So we may need minimum_maximum channels (like as your proposal) such as 2_12 for <IAC-Specific-needs>.
So, codecs parameter can be for example something like: aiac.<the order of Ambisonics>.<min channels_max channels>.mp4a.40.2
or aiac.<the order of Ambisonics>.<min channels_max channels>.Opus

**Comment by cconcolato** [comment]
In this case, I think the Ambisonics order and the min max values should be in the sample entry. 
Maybe min is always 2?

**Comment by sunghee-hwang** [comment]
Ok. I see. I think that min is the number of channels for BCG.

**Comment by cconcolato** [comment]
I think we agree to include the following in the `codecs` parameter:
- use of Ambisonic
- order of ambisonic
- use of non-diegetic streams
- channel counts (min and max) for non-diegetic streams

Looking again at all the [features of IAC](https://github.com/AOMediaCodec/iac/issues/21#issuecomment-1065379909), I wonder if the following shouldn't be also in the `codecs` parameter:
- use of ambisonics demixing?
- exact channel layout? For existing codecs, the layout is usually conveyed in DASH in the `ChannelConfiguration` descriptor. But one has to invent a new mapping for every non-DASH application. Having it in the `codecs` parameter would make it generic, no?

However, in my view the following do not need to be advertised (i.e. a receiver can handle all variations)
- demixing coefficients are not needed. It seems that all receivers should be able to handle all coefficients, no?
- the fact that coupling is used seems to be a basic feature that all receiver would support, no?

**Comment by sunghee-hwang** [comment]
> I think we agree to include the following in the `codecs` parameter:
> 
> * use of Ambisonic
> * order of ambisonic
> * use of non-diegetic streams
> * channel counts (min and max) for non-diegetic streams
> 
> Looking again at all the [features of IAC](https://github.com/AOMediaCodec/iac/issues/21#issuecomment-1065379909), I wonder if the following shouldn't be also in the `codecs` parameter:
> 
> * use of ambisonics demixing?
I will add it
> * exact channel layout? For existing codecs, the layout is usually conveyed in DASH in the `ChannelConfiguration` descriptor. But one has to invent a new mapping for every non-DASH application. Having it in the `codecs` parameter would make it generic, no?
> 
In this case, the min and max channel layout may be enough?? 

> However, in my view the following do not need to be advertised (i.e. a receiver can handle all variations)
> 
> * demixing coefficients are not needed. It seems that all receivers should be able to handle all coefficients, no?
Yes they should do
> * the fact that coupling is used seems to be a basic feature that all receiver would support, no?
What does the "coupling" mean?

**Comment by cconcolato** [comment]
> In this case, the min and max channel layout may be enough??

In your [slide 9](https://groups.aomedia.org/g/WG-Storage-and-Transport/files/Immersive_Audio_Container/AOM_Immersive3DAudio_20220207_r1.pdf), the min would be 1 (mono) and the max would be (12 = 7.1.4). It it obvious to the receiver that they can also reconstruct the intermediate 2, 3.1.2 and 5.1.2?

**Comment by sunghee-hwang** [comment]
> > In this case, the min and max channel layout may be enough??
> 
> In your [slide 9](https://groups.aomedia.org/g/WG-Storage-and-Transport/files/Immersive_Audio_Container/AOM_Immersive3DAudio_20220207_r1.pdf), the min would be 1 (mono) and the max would be (12 = 7.1.4). It it obvious to the receiver that they can also reconstruct the intermediate 2, 3.1.2 and 5.1.

Yes, right. For audio scalability, 7.1.4ch can be only reconstructed through reconstructing intermediate channel layouts.

**Comment by tdaede** [comment]
Would the channel count range be in the sample entry as well? Otherwise one would need to parse all the sample groups of a file to figure out the codecs parameter - not terrible but a bit unusual.

Also, AFAIK the client usually selects an appropriate audio track not by the codecs string but rather by e.g. DASH channel layout manifest information. Will we need to define that as well, and could that potentially substitute for that information in the codecs string?

**Comment by tdaede** [comment]
Codec negotiation via HTTP is one possible use case for the channel layout information being present in the codec string.

**Comment by tdaede** [comment]
We should determine if the channel layouts are an ordered list that we can specify only the min and max of, or if we need to explicitly list every possible channel layout we can decode to.

**Comment by sunghee-hwang** [comment]
> We should determine if the channel layouts are an ordered list that we can specify only the min and max of, or if we need to explicitly list every possible channel layout we can decode to.

Let me explain regarding an encoder guideline for IAC.
Below figure shows the allowed downmix pathes for audio scalability of channel audio.
![image](https://user-images.githubusercontent.com/97494915/159395873-d3c97998-119b-4a38-91b3-33bfeb163722.png)
As you can see, each path define its enc. So, it can be downmixed by using the combination of encs from base channel to target channel. Demixing is done by mirroring. So, if some decoders (which understand IAC) support max layout, then I think they can support the layouts which are downmixed from the max layout. I mean they already know from the max layout which layouts they can support.
Also if you refer to the below figure, then you can see the hierarchical downmixing is separately done between surround and height channels. Only one exception is to 3.1.2 downmix from 5.1.2ch which needs the bridge between them.
![image](https://user-images.githubusercontent.com/97494915/159396810-c6ca31d7-80a9-4328-8df6-d21380629693.png)

As summary, IMO we may not need every possible channel layout. If min and max layouts are not enough, then we may add one more layout for 3.1.2ch.

**Comment by tammy-lees** [comment]
Is it usually required to notify channel layouts in MIME `codecs` parameter? Otherwise, how about to simply start with the following: 
`aiac.<base-codecs-value>[.PP]` where `PP` is the value to specify the profile number if necessarily defined for legal range of channel layouts or use of ambisonics or use of scalable stream? In our case, `aiac.opus` or `aiac.mp4a.40.2` would be examples.
In consideration that the main use case of IAC would be 3D audio (which is capable of performing multi channels decoding and matrix process), I think that playable decision would be mainly dependent on which base codec is used for substream coding.

**Comment by cconcolato** [comment]
@tammy-lees In HTTP Streaming systems (DASH, HLS, ...), the client needs to decide what streams can download and among these streams what is the one that will give the best user experience, taking into account its capabilities. For audio, signaling the channel layout is part of it. 

In DASH, this is done with the `AudioChannelConfiguration` descriptor.  The [DASH-IF IOP](https://dashif.org/docs/IOP-Guidelines/DASH-IF-IOP-Part8-v5.0.0.pdf) indicates how to use them, but this is complicated. 

In [HLS](https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis), there is a way to indicate the number of channels (search for `CHANNELS`). 

In [W3C Media Capabilities](https://w3c.github.io/media-capabilities/#audioconfiguration), it's yet another way to do it.

In summary, there is no unified way. In my view, having layout information in the `codecs` parameter would make it much simpler for clients.

As for the construction of the `codecs` parameter, I don't understand what the "profile number" would be.  Also, I have a preference for putting the base codecs at the end, because you can simply pass the base codecs string to your application to know if they support it. You don't even have to parse it, just pass the end of the string after the IAC codecs parameter.

**Comment by tammy-lees** [comment]
I understand that there is no unified way. The question is whether IAC proposes to provide the unified way in codecs parameter for the first time. If we can find the simplest way, then it will be welcome as you recommend it.

Here are some issues we need to consider to support it:
(1) The number of channels can not specify the exact one channel layouts.
For example, 7.1 = 5.1.2 (# of channels = 8), 5.1 = 3.1.2 (# of channels = 6)
(2) min-max cannot specify supporting combinations exactly (with only demixing operations):
For example, given that 7.1.4 as max channel layouts and 5.1 as min channel layouts,
One possible combination is 5.1 - 7.1 - 7.1.4
The other may be 5.1 - 7.1.2 - 7.1.4

The reason why I consider profiles is to shorten codec strings.
For profiles,
we may divide supporting layout into
00: mono/stereo
01: surround
02: 3D channel layouts
03: ambisonics
04: 00~02 + scalable
05: reserved for future extension ?
In this case, it is still not sufficient to indicate the exact channel layouts to be supported.
If needed, we can define more information to specify exact channel layouts followed by a series of period (.) separated two-digit value for each profile.

However, most use cases of IAC would be 02 or more. So I am not sure that inclusion of profiles is efficient.
Please note that the above profiling is just one example to explain what I suggested.

**Comment by cconcolato** [comment]
I agree that giving only the channel count is not sufficient. It does not tell you in which branch of the downmix diagram you are. 10 channels could be on the left side (5.1.4) or the right side (7.1.2). Except when you give 12, there is only one choice.

IIUC, providing max-min let the decoder know what it cannot do. Do I understand correctly that given the 'max' layout, you can always produce all layout below? For example, you can always produce stereo or mono. I'm still trying to understand what 'min' tells you.

**Comment by tammy-lees** [comment]
Could you explain what IIUC stand for?

Regarding to your questions, we have to clarify how to produce target layouts.
In formative spec, demixing operations need to be defined to untangle mixed channels of the lower layer (CG) to be used as ingredients for the higher layer. So target layouts to be producible by formative action are pre-determined by encoders.
In informative spec (we can recommend some downmix matrices), producing from the higher configuration to the lower configuration can be possible by selecting appropriate downmix matrix on client/device side by their own choice.

So if an encoder sent a scalable stream of 5.1 - 7.1 - 7.1.4, then stereo is not producible because a decoder try to decode 5.1 first and do demixing for the next layer's reconstruction (7.1). However, if we provide downmix matrix to convert 5.1 to stereo in informative part or use one of existing methods, stereo actually can be producible. But I think that it is out of scope and it is upto client/device decision.

It is noted that because surround down-mix path (upward demix path in decoder side) and height down-mix path (upward demix path in decoder side) are separated in proposal (except one cross over case) and paths are needed to be gone through always between start(min) and end(max). So we can consider to define as Ss_Se, Hs_He (Ss: Surround start value, Se: Surround end value, Hs: Height start value, He: Height end value) to specify producible layouts. And producible layouts are 2D cases are 'Ss.1.0 to Se.1.0 (in case of Ss == 1/2, mono/stereo Ss.0.0 and not used for 3D) including intermediate nodes in surround path' and 3D cases are nodes in surround path x nodes in height path saying 'Ss.1.Hs to Se.1.He'. Here we need to assign different notion (different digits or different character) between TF2 and T2 in Hs or He.

So, for example,
(1) 5.1 - 7.1 - 7.1.4 case
Ss = 5, Se = 7, Hs = 4, He = 4
--> 2D producible layouts: 5.1.0, 7.1.0
--> 3D producible layouts: 5.1.4, 7.1.4 (even though 5.1.4 is not intended, but supported by applying only demixing op)
(2) 5.1 - 7.1.2 - 7.1.4 case
Ss = 5, Se = 7, Hs = 2, He = 4
--> 2D producible layouts: 5.1.0, 7.1.0
--> 3D producible layouts: 5.1.2, 7.1.2, 5.1.4, 7.1.4 (even though 5.1.4, 7.1 are not intended, but supported by appying only demixing op)
With this notion, it is also noted that producible layouts are not sync with CGs. For example of 5.1.4, all CGs (sent for 7.1.4) are parsed and decoded to produce 5.1.4. On the other hands, in case of 5.1.2 of (2), upto the second CG information (sent for 7.1.2) is needed at least.

**Comment by cconcolato** [comment]
We agree with the slides from April 11, with the following:
- adding the version field
- signaling of Ambisonics mode with 2 or 3 digits depending on the order
- signaling of the layers as intended by the author
We need to discuss if have variable length codecs parameter is a problem or not.

**Comment by tdaede** [comment]
Closing this issue. Issue #41 can be used to discuss the variable length codecs parameter.
